rust/agama-network/src/dbus/tree.rs

```diff
@@ -0,0 +1,206 @@
+        Ok(())
+    }
+
+    /// Clears all the connections from the tree.
+    async fn remove_connections(&self) -> Result<(), ServiceError> {
+        let mut objects = self.objects.lock();
+        for path in objects.connections.values() {
+            self.remove_connection_on(path.as_str()).await?;
+        }
+        objects.connections.clear();
+        Ok(())
+    }
+
+    /// Refreshes the list of connections.
+    ///
+    /// TODO: re-creating the tree is kind of brute-force and it sends signals about
+    /// adding/removing interfaces. We should add/update/delete objects as needed.
+    pub async fn refresh_connections(
+        &self,
+        connections: &Vec<Connection>,
+    ) -> Result<(), ServiceError> {
+        self.remove_connections().await?;
+        self.add_connections(connections).await?;
+        Ok(())
+    }
+
+    /// Removes a connection object on the given path
+    ///
+    /// * `path`: connection D-Bus path.
+    async fn remove_connection_on(&self, path: &str) -> Result<(), ServiceError> {
+        let object_server = self.connection.object_server();
+        _ = object_server.remove::<interfaces::Wireless, _>(path).await;
+        object_server.remove::<interfaces::Ipv4, _>(path).await?;
+        object_server
+            .remove::<interfaces::Connection, _>(path)
+            .await?;
+        Ok(())
+    }
+
+    async fn add_interface<T>(&self, path: &str, iface: T) -> Result<bool, ServiceError>
+    where
+        T: zbus::Interface,
+    {
+        let object_server = self.connection.object_server();
+        Ok(object_server.at(path.clone(), iface).await?)
+    }
+}
+
+/// Objects paths for known devices and connections
+#[derive(Debug, Default)]
+pub struct ObjectsRegistry {
+    pub devices: HashMap<String, String>,
+    pub connections: HashMap<Uuid, String>,
+}
+
+impl ObjectsRegistry {
+    /// Registers a network device.
+    ///
+    /// * `name`: device name.
+    /// * `path`: object path.
+    pub fn register_device(&mut self, name: &str, path: &str) {
+        self.devices.insert(name.to_string(), path.to_string());
+    }
+
+    /// Registers a network connection.
+    ///
+    /// * `uuid`: connection UUID.
+    /// * `path`: object path.
+    pub fn register_connection(&mut self, uuid: Uuid, path: &str) {
+        self.connections.insert(uuid, path.to_string());
+    }
+
+    /// Returns the path for a connection.
+    ///
+    /// * `uuid`: connection UUID.
+    pub fn connection_path(&self, uuid: Uuid) -> Option<&str> {
+        self.connections.get(&uuid).map(|p| p.as_str())
+    }
+
+    /// Deregisters a network connection.
+    ///
+    /// * `uuid`: connection UUID.
+    pub fn deregister_connection(&mut self, uuid: Uuid) -> Option<String> {
+        self.connections.remove(&uuid)
+    }
+
+    /// Returns all devices paths.
+    pub fn devices_paths(&self) -> Vec<String> {
+        self.devices.values().map(|p| p.to_string()).collect()
```

Comment:
just matter of taste, but as values is already String, I usually prefer `to_owned` over `to_string` as it is not type conversion.